### PR TITLE
Refactor memory checking

### PR DESCRIPTION
**Pull request by moodlezoup**

### CODE REVIEW

File: jolt-core/src/jolt/vm/pc.rs [hunk 1]
```diff
@@ -136,25 +118,94 @@ pub struct PCPolys<F: PrimeField> {
...
+impl<F, G> StructuredPolynomials for PCPolys<F, G>
+where
+  F: PrimeField,
+  G: CurveGroup<ScalarField = F>,
+{
+  type Commitment = ProgramCommitment<G>;
+  type BatchedPolynomials = BatchedPCPolys<F>;
+
+  fn batch(&self) -> Self::BatchedPolynomials {
+    // TODO(JOLT-48): DensePolynomial::merge should be parameterized with  Vec<&DensePolynomial> given it is read only. Avoids clones.
```
[the remark this answers is not in the document]
Comment (sragss) [reply jolt-core/src/jolt/vm/pc.rs]: Done.

File: jolt-core/src/jolt/vm/pc.rs [hunk 2]
```diff
@@ -136,25 +118,94 @@ pub struct PCPolys<F: PrimeField> {
...
-	// to the nearest power of 2
+    // Note: a_read_write and read_cts have been implicitly padded with 0s
+    // to the nearest power of 2
     let mut a_read_write_usize: Vec<usize> = vec![0; num_ops];
     let mut read_cts: Vec<usize> = vec![0; num_ops];
     let mut final_cts: Vec<usize> = vec![0; code_size];
 
-	// TODO(sragss): Current padding strategy doesn't work. As it adds phantom
-	// reads, but no corresponding writes to final.
+    // TODO(sragss): Current padding strategy doesn't work. As it adds phantom
```
[the remark this answers is not in the document]
Comment (sragss) [reply jolt-core/src/jolt/vm/pc.rs:206]: Done.

File: jolt-core/src/lasso/memory_checking.rs [hunk 3]
```diff
@@ -1,698 +1,833 @@
...
+        hash_read: read_hashes[i],
+        hash_write: write_hashes[i],
+      };
+      debug_assert_eq!(
+        hashes.hash_init * hashes.hash_write,
+        hashes.hash_final * hashes.hash_read,
+        "Multiset hashes don't match"
+      );
+      multiset_hashes.push(hashes);
+      // TODO: append to transcript
```
[the remark this answers is not in the document]
Comment (sragss) [reply jolt-core/src/lasso/memory_checking.rs]: Done.

File: jolt-core/src/lasso/memory_checking.rs [hunk 4]
```diff
@@ -1,698 +1,833 @@
...
       // Multiset equality check
       assert_eq!(
-        eval.hash_init * eval.hash_write,
-        eval.hash_read * eval.hash_final
+        hash.hash_init * hash.hash_write,
+        hash.hash_read * hash.hash_final
       );
-
-      eval.append_to_transcript::<G>(transcript);
+      // TODO: append to transcript
```
[the remark this answers is not in the document]
Comment (sragss) [reply jolt-core/src/lasso/memory_checking.rs]: Done.

File: jolt-core/src/jolt/vm/mod.rs [hunk 5]
```diff
@@ -245,578 +103,31 @@ pub trait Jolt<F: PrimeField, G: CurveGroup<ScalarField = F>> {
...
-    commitments
-      .E_commitment
-      .append_to_transcript(b"comm_poly_row_col_ops_val", transcript);
-
-    let eq = EqPolynomial::new(r.to_vec());
-    let sumcheck_claim = Self::compute_sumcheck_claim(&ops, &polynomials.E_polys, &eq);
-
-    // TODO(sragss): rm
-    println!("Jolt::vm::prove() compute_sumcheck_claim result: {sumcheck_claim:?}");
+  fn prove_memory(memory_trace: Vec<MemoryOp>, memory_size: usize, transcript: &mut Transcript) {
```
[the remark this answers is not in the document]
Comment (moodlezoup) [reply jolt-core/src/jolt/vm/mod.rs:106]: yes exactly, see the below `todo!`

File: jolt-core/src/jolt/vm/instruction_lookups.rs [hunk 6]
```diff
@@ -0,0 +1,1245 @@
...
+            )
+          })
+          .map(|tuple| Self::fingerprint(&tuple, gamma, tau))
+          .collect();
+        DensePolynomial::new(leaf_fingerprints)
+      })
+      .collect()
+  }
+
+  fn read_write_grand_product(
```
Comment (sragss) [inline jolt-core/src/jolt/vm/instruction_lookups.rs:550]: nit: Note that this is an override.

File: jolt-core/src/lasso/surge.rs [hunk 7]
```diff
@@ -1,579 +1,742 @@
...
-      t * gamma.square() + v * *gamma + a - tau
-    };
+pub struct SurgeFinalOpenings<F, G>
+where
+  F: PrimeField,
+  G: CurveGroup<ScalarField = F>,
+{
+  final_openings: Vec<F>, // C-sized
+  final_opening_proof: CombinedTableEvalProof<G>,
+  a_init_final: Option<F>,      // Computed by verifier
```
[the remark this answers is not in the document]
Comment (sragss) [reply jolt-core/src/lasso/surge.rs:269]: Options as they're filled in by the verifier during `compute_verifier_openings`. Should add to the doc comment for `compute_verifier_openings`

File: jolt-core/src/poly/dense_mlpoly.rs [hunk 8]
```diff
@@ -266,6 +268,18 @@ impl<F: PrimeField> DensePolynomial<F> {
     DensePolynomial::new(Z)
   }
 
+  pub fn combined_commit<G>(
```
[the remark this answers is not in the document]
Comment (sragss) [reply jolt-core/src/poly/dense_mlpoly.rs:271]: Yes.

**Pull request merged**